metagpt/tools/web_browser_engine.py

```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env python
+
+from __future__ import annotations
+import asyncio
+import importlib
+
+from typing import Any, Callable, Coroutine, overload
+
+from metagpt.config import Config
+from metagpt.tools import WebBrowserEngineType
+from bs4 import BeautifulSoup
+
+
+class WebBrowserEngine:
+    def __init__(
+        self,
+        engine: WebBrowserEngineType | None = None,
+        run_func: Callable[..., Coroutine[Any, Any, str | list[str]]] | None = None,
+    ):
+        self.config = Config()
+        engine = engine or self.config.web_browser_engine
+
+        if engine == WebBrowserEngineType.PLAYWRIGHT:
+            web_browser_engine = importlib.import_module("metagpt.tools.web_browser_engine_playwright")
+            run_func = web_browser_engine.PlaywrightWrapper().run
+        elif engine == WebBrowserEngineType.SELENIUM:
+            web_browser_engine = importlib.import_module("metagpt.tools.web_browser_engine_selenium")
+            run_func = web_browser_engine.SeleniumWrapper().run
+        elif engine == WebBrowserEngineType.CUSTOM_ENGINE:
+            run_func = run_func
+        else:
+            raise NotImplementedError
+        self.run_func = run_func
+        self.engine = engine
+
+    @overload
+    async def run(self, url: str) -> str:
+        ...
+
+    @overload
+    async def run(self, url: str, *urls: str) -> list[str]:
+        ...
+
+    async def run(self, url: str, *urls: str) -> str | list[str]:
+        page = await self.run_func(url, *urls)
+        if isinstance(page, str):
+            return get_page_content(page)
+        return [get_page_content(i) for i in page]
+
+
+def get_page_content(page: str):
+    soup = BeautifulSoup(page, "html.parser")
+    return "\n".join(i.text.strip() for i in soup.find_all(["h1", "h2", "h3", "h4", "h5", "p", "pre"]))
```

Comment:
It seems that can only get the content only with the html page meets the templates. What if the user want to parse html page with own regex or templates?

Reply:
[add parse_func for WebBrowserEngin to support custom parsing](https://github.com/geekan/MetaGPT/pull/69/commits/0700cd2e62477227034017ea3fd8bc65f98446cb)